_posts/2022-05-16-opentelemetry-metrics-for-data-prepper.md

```diff
@@ -0,0 +1,114 @@
+
+## Why Ingest Metrics in OpenSearch?
+
+At a first glance, OpenSearch offers great support for storing and analyzing any structured observability data.
+It provides powerful search and aggregation capabilities.
+Excellent visualizations support users in generating insights.
+The Vega visualizations lift the metrics use-case to a whole new level.
+OpenSearch already comes with alerting and anomaly detection, that nicely extends to metrics.
+
+However, OpenSearch was not built as a time series database.
+Storing metrics will be a lot less efficient with regard to resource consumption in comparison to specialized solutions.
+But, OpenSearch can shine on special use-cases: High-cardinality data sets.
+Metrics at least consists of a name, a timestamp and a value.
+Additionally, attributes can be attached that describe the nature or origin of the metrics better.
+OpenSearch does not care about the number of different values in any of those fields, specialized time series databases often do.
+Let's say you count the number of service request by a service key.
+Depending on the number of service keys, this might be problematic.
+
+When storing metrics in OpenSearch you need to be mindful of the overall event rate, but you do not need to plan ahead for cardinality.
+You will be able to augment logs and traces by a careful selection of metrics to enhance your overall observability.
+
+## Solution Architecture
+
+Data Prepper now supports metrics ingestion into OpenSearch using the OpenTelemetry protocol.
+There are many implementations capable of sending metrics using this protocol.
+The OpenTelemetry Collector was chosen as an agent for metrics acquisition in the following diagram.
+It can pull several metrics endpoints or act as a receiver for others.
+Metrics are sent to Data Prepper using the gRPC-based OpenTelemetry protocol.
+Data Prepper receives the metrics, dissects and maps their data points and saves each data point as an individual OpenSearch document.
+The diagram outlines the basic architecture for metrics ingestion using OpenTelemetry collector, Data Prepper and OpenSearch.
+
+![Architecture]({{ site.baseurl }}/assets/media/blog-images/2022-05-16-opentelemetry-metrics-for-data-prepper/architecture.drawio.png){: .img-fluid}
+
+The OpenTelemetry metrics data model is described in great detail in its [specification.](https://opentelemetry.io/docs/reference/specification/metrics/datamodel/)
+Very briefly, metrics are divided into 5 different types:
+
+* Sum
+* Gauge
+* Histogram
+* ExponentialHistogram
+* Summary
+
+Each metric type has a specific data point format.
+For example, sums and gauges use single value data points, while histograms and summaries used nested arrays to represent buckets of data.
+
+Correlation between different metrics as well as traces and logs is provided by attributes and trace/span ids.
+Direct connections to spans can be included with the Exemplars feature of OpenTelemetry.
```

Comment:
Can you elaborate on this somewhat? For example, how do the exemplars provide a connection to spans? I think just a high-level comment about this relationship would help.

Reply:
Support for exemplars is in the same PR as the exponential histograms. If that does not make it, I will remove this sentence entirely, since the feature will not be supported, yet.